### PR TITLE
Support setting cargo features and resolve `default` features by default

### LINKED ISSUE #2524: no more type annotations displayed
While working on [cargo-cache](https://github.com/matthiaskrgr/cargo-cache) I noticed that RA stopped displaying any type annotations.

When I start a new cargo project and write `let x = "hello";`, RA correctly displays `let x: &str = "hello";` but inside `cargo-cache`, it no longer does that for some reason.
I tried adding the line to random places without success (type annotations were missing). :/ 

RA @ 5202b0ecbaab2d44e886828f0e30dee67fad0c5c

### PR DESCRIPTION
**Pull request by oxalica**
Fixes #2524

### CODE REVIEW

File: editors/code/package.json [hunk 1]
```diff
@@ -278,6 +278,21 @@
...
                     "description": "Maximum length for inlay hints"
+                },
+                "rust-analyzer.cargoFeatures.noDefaultFeatures": {
+                    "type": "boolean",
+                    "default": false,
+                    "description": "Do not activate the `default` feature"
+                },
+                "rust-analyzer.cargoFeatures.allFeatures": {
+                    "type": "boolean",
+                    "default": false,
```
Comment (matklad) [inline editors/code/package.json]:
```suggestion
                    "default": true,
```

I think that features a *usually* additive, so by default `allFeatures` make slightly more sense for our use-case.

**Comment by matklad** [comment]
Excellent, this is super useful! Let's add docs for this to https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/user/README.md#settings as well?

**Comment by flodiebold** [comment]
Are the plugin settings the right place to configure desired features? It seems a bit weird to me, but I'm not familiar with how VSCode handles them.

Alternatively, would it be possible to configure the desired features in Cargo.toml? I.e. if there's a `rust-analyzer` feature, we just take that.

**Comment by matklad** [comment]
I definitely think that Cargo.toml is not the right place -- you shouldn't edit `Cargo.toml` if you want to switch from one feature set to another in an IDE.

Ie, I think switching should be dynamic (preferably, with a dedicated UI which allows toggling feature per package, with autocompletion), and that it should be persisted between editor sessions. 

More generally, I feel there's a missing feature in the LSP spec: it should specify some kind of `lsp.json` file to put language server configs into, which is common between the editors. In the current situation, one has to use either per-editor config, or per-langauge server config (`rust-analyzer.toml`), neither of which is ideal. And there's a problem of where to put docs for the config...

I am not a fan of implementing `rust-analyzer.toml` or some such, as it seems tricky to get really useful. For example, vscode automatically merges global config in `~/.config/vscode` and project-config in `/proejct-dir/.vscode`.

**Comment by lnicola** [comment]
We can add custom `Cargo.toml` metadata, e.g. https://github.com/serde-rs/serde/blob/master/serde/Cargo.toml#L26-L27.

**Comment by matklad** [comment]
@lnicola we can, but it doesn't change the fact that "currently active set of features" is a dynamic value, which user might want to change frequently. 

And I think metadata section won't work well for more static configs, like "insert ()" -- those should be per-user and overridable, and not per-project.

**Comment by lnicola** [comment]
I know. I was suggesting that as an alternative to a `rust-analyzer` feature. I can imagine, though, that some project might want to say "feature X is totally broken under RA, let us disable it". But that's more of an edge case, and we probably shouldn't optimize for it unless people start asking.

**Comment by kjeremy** [comment]
The right thing to do would probably be to implement [workspace/configuration](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#workspace_configuration) which would get triggered from vscode settings changes. I started down that path but never finished it.

**Comment by matklad** [comment]
Thanks!

bors r+

not sure what's up with CI

**Comment by ice1000** [comment]
> Thanks!
> bors r+
> not sure what's up with CI

Should be a bug of GitHub Actions

**Comment by edwin0cheng** [comment]
bors retry

**Pull request merged**